buildtools/npm/lockfile.go

```diff
@@ -0,0 +1,125 @@
+package npm
+
+import (
+	"path/filepath"
+
+	"github.com/fossas/fossa-cli/errors"
+	"github.com/fossas/fossa-cli/files"
+	"github.com/fossas/fossa-cli/graph"
+	"github.com/fossas/fossa-cli/pkg"
+)
+
+var PossibleLockfileFilenames = [...]string{"package-lock.json", "npm-shrinkwrap.json"}
+
+type Lockfile struct {
+	Name         string
+	Version      string
+	Dependencies Dependencies
+}
+
+type DependencyLockEntry struct {
+	Version      string
+	Requires     map[string]string
+	Dependencies map[string]*DependencyLockEntry
+}
+
+type Dependencies map[string]*DependencyLockEntry
+type DependencyMap map[pkg.ID][]pkg.ID
+type DeepDepIdsSet map[pkg.ID]bool
+
+// FindAndReadLockfile checks the root of the node project at the given path for lockfiles, then returns the parsed
+// contents if one is found
+func FindAndReadLockfile(path string) (Lockfile, error) {
+	var lockfile Lockfile
+
+	for _, filename := range PossibleLockfileFilenames {
+		filePath := filepath.Join(path, filename)
+		exists, err := files.Exists(filePath)
+
+		if err == nil && exists {
+			err := files.ReadJSON(&lockfile, filePath)
+			return lockfile, err
+		}
+	}
+
+	return lockfile, errors.Newf("none of %v found at root of node project", PossibleLockfileFilenames)
+}
+
+// scanLockfileForDependencyInfo recursively scans a lockfile to construct a map of package IDs to their requirements
+// as well as a set of the package IDs that have been required by other dependencies somewhere else in the file
+func scanLockfileForDependencyInfo(deps Dependencies) (DependencyMap, DeepDepIdsSet) {
```

Comment:
[nit] I think the function name here is a little unclear since the argument is type `Dependencies`. Something like `retrieveTransitiveInformation` might be a little more clear, but I'm not sure.

Reply:
Yeah, it's a weird function so naming is tough. I think `retrieveTransitiveInformation` makes more sense 🙂